### ISSUE TITLE
NTPD time_offset-loop is wrong

### ISSUE THREAD

**Issue by PierreF** [collectd/collectd#1300]
With collectd 5.4 and 5.5, the time_offset-loop does not match value from ntpq/ntpdc.

For example, right now:

```
# I'm using graphite output. Here is ntpd releated measure from REQ_GET_KERNINFO
hostname.ntpd.frequency_offset-loop -6.302734375 1444644419
hostname.ntpd.time_offset-loop 86.6073455810547 1444644419
hostname.ntpd.time_offset-error 0.201217651367188 1444644419

# And output of ntpdc -c kerninfo run at the same time
pll offset:           0.0056759 s
pll frequency:        -6.303 ppm
maximum error:        0.421666 s
estimated error:      0.013187 s
status:               2001  pll nano
pll time constant:    8
precision:            1e-09 s
frequency tolerance:  500 ppm
```

After few digging in both collectd and ntpdc, It seem that NTPD changed it's factor from 65536.0 to 1 000 000, which explain the error:

86.6073455810547 (collectd value of offset-loop) * 65536.0 / 1 000 000 = 5.675899000000001

ntpdc factor 1 000 000 (us) comes from ntpdc/ntpdc_ops.c file, function kerninfo. Note that in this function, the scale factor could be 1 000 000 000 (ns) if STA_NANO is defined :(

I'm using ntp version 4.2.6.p5+dfsg-3ubuntu6

**Comment by mfournier** [comment]
Thanks for tracking this down @PierreF !

So if I'm following correctly, ntpd changed this multiplication factor at some point, right ? In this case, your patch would cause the opposite problem for folks stiil using older ntpd versions, wouldn't it ?

I see an `INFO_VERSION()` macro in `src/ntpd.c`, would this be the ntpd version by any chance ? Would it be an idea to have this multiplication factor vary depending on this version ?

**Comment by PierreF** [comment]
It was what I thought, that ntpd changed this multiplication factor... but the oldest ntpd version I can found (4.1.1, dated of ~ Febuary 2002) already use that factor.

Since ntp.c was added in collectd in 2006 (based on git log ntp.c), I think in fact it never worked.

I've confirmed this using Debian 4.0 in a VM, I've the same offset (8 ms for ntpdc -c kerninfo / 124 ms from collectd).

Note that only offset and estimated error from kerninfo is affected. Not the offset for a given peer. I suppose most user only look at offset from peer, that maybe why nobody noticed this earlier.


Version used by Debian 4.0:

ntp 4.2.2 (dated from 22 November 2009 - based on Debian changelog)
collectd 3.10.4

**Comment by mfournier** [comment]
Yes, this seems reasonable to me (@PierreF just pinged me about this on IRC), so I'd +1 this fix.

@octo any thoughts about this, as you have authored these lines ? Would this be deliberate ? It doesn't look like it is, wrt/ code comments & docs.

**Comment by octo** [comment]
LGTM. It sounds likely I got the factors mixed up.

**Comment by mfournier** [comment]
Closed by 53438c510. The fix for this issue will be part of collectd 5.4.3, and 5.5.1. Thanks !

**Comment by jcberthon** [comment]
I got bitten by this exact issue on Debian Jessie which ships with collectd 5.4.1-6+deb8u1.

I'm not using the peer offset, because with collectd there is no way to know which peer was selected by ntp for synchronisation (the '*' in the peer status returned by ntpq). And I think, IMHO, that the system offset is a better measure of the accuracy of the local clock with respect to the UTC (which hopefully is what the remote peers are synchronised to!).

However, what offset the kerninfo command returns is not the same as the one the rv (displaying system variable) returns. And as far as I understood the NTPv4 RFC, the latter provide the system variables which I think provides the best evaluation of the NTP performance. To quote the RFC section 11.2.3:

    The system variables are available to dependent application programs
    as nominal performance statistics.  The system offset THETA is the
    clock offset relative to the available synchronization sources.  The
    system jitter PSI is an estimate of the error in determining this
    value, elsewhere called the expected error.

_Note: the system variables mentioned are the ones returns by `ntpq -c rv`, specifically the offset and sys_jitter. This can be retrieved using the command line: `ntpq -c "rv 0 offset,sys_jitter"`. The offset returned by the `ntpdc -c loopinfo` command seems to be the same as the one returned by the `ntpq -c rv` command._

Here is a comparison of both outputs. I'm gathering statistics in InfluxDB. One producer is collectd 5.6.1 (running on Alpine Linux in Docker), it is displayed in yellow below. And the other is a small BASH script which by combining `ntpq -c "rv 0 offset"` command, some awk and curl is pushing the system offset data to the same InfluxDB database, it is displayed in green below. The monitored ntpd server is running 4.2.8p4 on Ubuntu 16.04 in Docker, it is synchronised using the pool servers provided by the NTP pool project. The graphic below was produced using Grafana (for collect):

![2016-11-28 14_03_30-grafana - ntpv2](https://cloud.githubusercontent.com/assets/1397088/20669394/3cca0ff0-b574-11e6-8ec4-09a1af5216ff.png)

_Note: the weird thing with the offset - as monitored by collectd - is that the value is always positive, eventhough running `ntpdc -c kerninfo` shows sometime negative values. This is not too important but worth mentioning._

_Note2: for information here is the grafana query used to get the offset from collectd:_

    SELECT mean("value") FROM "ntpd_value" WHERE "host" = 'myhost' AND "type" = 'time_offset' AND "type_instance" = 'loop' AND $timeFilter GROUP BY time($interval) fill(null)
